### ISSUE TITLE
将多区域 loss_mask 高级配置代码集成到 verl 项目中

### ISSUE THREAD

**Issue by WillQvQ** [WillQvQ/arXiv-MCP-Demo#1]
# 将多区域 loss_mask 高级配置代码集成到 verl 项目中

## 概述

根据 Issue #5 中提出的多区域 loss_mask 解决方案，本 issue 详细说明如何将相关代码集成到 verl 项目的现有架构中。

### 1. 核心数据集类

**文件位置**: `verl/utils/dataset/multi_region_sft_dataset.py`

**代码调用链分析**:
```
verl/trainer/fsdp_sft_trainer.py (第634行)
└── create_sft_dataset() 函数
    └── load_extern_type() 加载自定义数据集类
        └── MultiRegionLossMaskDataset (新增)
```

**集成方式**:
- 继承现有的 `SFTDataset` 类
- 与现有的 `MultiTurnSFTDataset` 并列
- 通过 `data.custom_cls` 配置项引用

### 2. 配置文件

**文件位置**: `verl/trainer/config/` 目录下新增三个配置文件：

- `multi_region_position_based.yaml`
- `multi_region_tag_based.yaml` 
- `multi_region_structured.yaml`

**调用链分析**:
```
main() 函数 (hydra 入口)
└── @hydra.main(config_path="config", config_name="配置名")
    └── 加载对应的 yaml 配置文件
        └── 传递给 FSDPSFTTrainer
```

### 3. 数据预处理工具

**文件位置**: `verl/utils/dataset/prepare_multi_region_data.py`

**用途**: 独立的工具脚本，用于生成示例数据和验证数据格式

### 4. 可视化和监控工具

**文件位置**: 
- `verl/utils/debug/multi_region_visualizer.py`
- `verl/utils/debug/training_monitor.py`

**调用链分析**:
```
FSDPSFTTrainer.training_step() (fsdp_sft_trainer.py 第278行)
└── _compute_loss_and_backward()
    └── 可选集成监控工具
```

### 5. 高级功能模块

**文件位置**:
- `verl/utils/dataset/dynamic_weight_adjuster.py`
- `verl/utils/dataset/region_dependency_handler.py`

### 步骤 1: 创建核心数据集类

在 `verl/utils/dataset/multi_region_sft_dataset.py` 中创建 `MultiRegionLossMaskDataset` 类：

```python
# 文件: verl/utils/dataset/multi_region_sft_dataset.py
import torch
import re
from typing import List, Dict, Any, Tuple
from verl.utils.dataset import SFTDataset

class MultiRegionLossMaskDataset(SFTDataset):
    def __init__(self, parquet_files, tokenizer, config):
        super().__init__(parquet_files, tokenizer, config)
        
        # 多区域配置
        self.region_config = config.get("multi_region_config", {})
        self.data_format = self.region_config.get("format", "position_based")
        self.max_regions = self.region_config.get("max_regions", 100)
        self.weight_normalize = self.region_config.get("weight_normalize", False)
        
        # 处理器映射
        self.processors = {
            "position_based": self._process_position_based,
            "tag_based": self._process_tag_based,
            "structured": self._process_structured
        }
    
    # [完整的类实现代码如 Issue #5 所述]
```

**关键集成点**:
- 在 `verl/utils/dataset/__init__.py` 中添加导入
- 修改 `create_sft_dataset()` 函数以支持新的数据集类型

### 步骤 2: 修改现有的 create_sft_dataset 函数

在 `verl/trainer/fsdp_sft_trainer.py` 第634行的 `create_sft_dataset()` 函数中添加对多区域数据集的支持：

```python
def create_sft_dataset(data_paths, data_config, tokenizer):
    """Create a dataset."""
    # 检查是否使用多区域配置
    if "multi_region_config" in data_config:
        from verl.utils.dataset.multi_region_sft_dataset import MultiRegionLossMaskDataset
        dataset_cls = MultiRegionLossMaskDataset
    # 检查自定义数据集类
    elif data_config.custom_cls.get("path", None):
        from verl.utils.import_utils import load_extern_type
        dataset_cls = load_extern_type(data_config.custom_cls.path, data_config.custom_cls.name)
    # 检查多轮对话数据集
    elif data_config.get("multiturn", {}).get("enable", False):
        dataset_cls = MultiTurnSFTDataset
    # 默认单轮数据集
    else:
        dataset_cls = SFTDataset

    dataset = dataset_cls(parquet_files=data_paths, tokenizer=tokenizer, config=data_config)
    return dataset
```

### 步骤 3: 创建配置文件

在 `verl/trainer/config/` 目录下创建三个新的配置文件：

**文件**: `verl/trainer/config/multi_region_position_based.yaml`
```yaml
# 基于位置标记的多区域配置
defaults:
  - sft_trainer
  - _self_

data:
  multi_region_config:
    format: "position_based"
    max_regions: 50
    weight_normalize: true
  max_length: 4096
  micro_batch_size_per_gpu: 2

model:
  enable_gradient_checkpointing: true

trainer:
  experiment_name: multi-region-position-based
```

**文件**: `verl/trainer/config/multi_region_tag_based.yaml`
**文件**: `verl/trainer/config/multi_region_structured.yaml`
（类似结构，内容对应不同格式）

### 步骤 4: 修改 loss 计算逻辑

在 `verl/trainer/fsdp_sft_trainer.py` 的 `_compute_loss_and_backward()` 方法中（第299行附近），确保正确处理多区域的 loss_mask：

```python
def _compute_loss_and_backward(self, batch, do_backward=True):
    """Compute loss with optional sequence parallelism and remove padding features"""
    # ... 现有代码 ...
    
    # 获取 loss_mask（支持多区域）
    loss_mask = batch.pop("loss_mask")[:, :-1].reshape(-1).to(self.device_name)
    
    # 如果有区域信息，可以进行额外处理
    if "region_info" in batch:
        region_info = batch.pop("region_info")
        # 可选：应用动态权重调整
        if hasattr(self.config.data, "multi_region_config"):
            # 集成动态权重调整逻辑
            pass
    
    # ... 现有的 loss 计算逻辑 ...
```

### 步骤 5: 添加调试和可视化工具

**文件**: `verl/utils/debug/multi_region_visualizer.py`

创建可视化工具类，用于调试和分析多区域 loss_mask 的效果。

### 步骤 6: 更新文档和示例

在 `examples/sft/` 目录下创建多区域训练的示例脚本：

**文件**: `examples/sft/multi_region/run_multi_region_position.sh`
```bash
#!/bin/bash
set -x

nproc_per_node=$1
save_path=$2
shift 2

torchrun --standalone --nnodes=1 --nproc_per_node=$nproc_per_node \
     -m verl.trainer.fsdp_sft_trainer \
    --config-name=multi_region_position_based \
    data.train_files=$HOME/data/multi_region/train.parquet \
    data.val_files=$HOME/data/multi_region/test.parquet \
    trainer.default_local_dir=$save_path \
    trainer.project_name=multi-region-sft \
    trainer.logger=['console','wandb'] \
    $@
```

### 单元测试

在 `tests/` 目录下创建测试文件：

**文件**: `tests/test_multi_region_dataset.py`
```python
import pytest
import torch
from verl.utils.dataset.multi_region_sft_dataset import MultiRegionLossMaskDataset

def test_position_based_processing():
    # 测试基于位置的数据处理
    pass

def test_tag_based_processing():
    # 测试基于标签的数据处理
    pass

def test_structured_processing():
    # 测试结构化数据处理
    pass
```

### 集成测试

创建端到端的训练测试：

```bash
# 测试多区域训练是否正常工作
python -m pytest tests/test_multi_region_integration.py -v
```

## 向后兼容性

- 所有新功能都是可选的，不影响现有的 SFT 训练流程
- 通过配置文件控制是否启用多区域功能
- 现有的 `SFTDataset` 和 `MultiTurnSFTDataset` 保持不变

## 性能考虑

- 多区域处理可能增加数据加载时间
- 长序列训练需要更多内存
- 建议在配置中启用梯度检查点和适当的并行策略

## 文件清单

新增/修改的文件列表：

### 新增文件
- `verl/utils/dataset/multi_region_sft_dataset.py`
- `verl/utils/dataset/prepare_multi_region_data.py`
- `verl/utils/debug/multi_region_visualizer.py`
- `verl/utils/debug/training_monitor.py`
- `verl/trainer/config/multi_region_position_based.yaml`
- `verl/trainer/config/multi_region_tag_based.yaml`
- `verl/trainer/config/multi_region_structured.yaml`
- `examples/sft/multi_region/run_multi_region_position.sh`
- `examples/sft/multi_region/run_multi_region_tag.sh`
- `examples/sft/multi_region/run_multi_region_structured.sh`

### 修改文件
- `verl/trainer/fsdp_sft_trainer.py` (第634行 create_sft_dataset 函数)
- `verl/utils/dataset/__init__.py` (添加新数据集类的导入)

## 总结

通过以上集成步骤，多区域 loss_mask 功能将无缝集成到 verl 项目中，为复杂长序列的精细化训练提供强大支持，同时保持与现有功能的完全兼容性。
